### PR TITLE
additional checking for CfIds on createTable

### PR DESCRIPTION
**Pull request by clockfort**
I haven't tested this yet.

I just want to confer with TomP and iterate with him here, before maybe merging this.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/739)

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java [hunk 1]
```diff
@@ -378,6 +379,31 @@ public InetSocketAddress apply(String endpoint) {
...
+
+            for (Entry<InetSocketAddress, CassandraClientPoolingContainer> poolForHost : currentPools.entrySet()) {
+                try {
+                    returnValuesByHost.put(poolForHost.getKey(), poolForHost.getValue().runWithPooledResource(f));
+                } catch (Exception e) {
+                    exceptionsByHost.put(poolForHost.getKey(), e);
+                }
+            }
+
+        ImmutableSet<V> returnValueSet = ImmutableSet.copyOf(returnValuesByHost.values());
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java]: this assumes that the values implement hashCode/equals properly, which is fine for case of CF ID `Integer`, but probably needs to be a warning on the method javadoc

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java [hunk 2]
```diff
@@ -1348,13 +1351,27 @@ private void createTablesWithLock(final Map<TableReference, byte[]> tableNamesTo
...
                 if (!tablesToCreate.isEmpty()) {
                     CassandraKeyValueServices.waitForSchemaVersions(client, "(all tables in a call to createTables)", configManager.getConfig().schemaMutationTimeoutMillis());
+                    for (Entry<CfDef, Integer> cfAndCfId : cfToOurCfId.entrySet()) {
+                        String cfName = cfAndCfId.getKey().getName();
+                        int consistentIdAroundCluster = CassandraKeyValueServices.checkCfIdOnAllHosts(clientPool, keyspace, cfName);
+                        if (consistentIdAroundCluster != cfAndCfId.getValue()) {
+                            log.error("Tried to create a table " + cfName + "; we expected its CfId to be " + cfAndCfId.getValue() +
+                                    " but it looks like someone else created it successfully and fully consistently around the cluster before we did with CfId " + consistentIdAroundCluster);
+                        }
+                    }
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java]: Should we do similar consistency check anywhere we `system_update_column_family` (`alterGcAndTombstone` and `internalPutMetadataForTables`)

**Comment by schlosna** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java, line 1333 \[r1\]](https://reviewable.io:443/reviews/palantir/atlasdb/739#-KNm_6RDRkN-Rx_xMEcb:-KNm_6RDRkN-Rx_xMEcc:323584404) ([raw file](https://github.com/palantir/atlasdb/blob/c817cd3402bb63a1dfd8bb9da84f0a24338b3831/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java#L1333)):*
> ```Java
>     public void createTables(final Map<TableReference, byte[]> tableNamesToTableMetadata) {
>         schemaMutationLock.runWithLock(() -> createTablesWithLock(tableNamesToTableMetadata));
>         internalPutMetadataForTables(tableNamesToTableMetadata, false);
> ```

Should we get rid of the `possiblyNeedToPerformSettingsChanges` argument to `internalPutMetadataForTables` so that we also acquire schema mutation lock, and assert that we hold `schemaMutationLock` anywhere we touch key space or column family?

```
public void createTables(final Map<TableReference, byte[]> tableNamesToTableMetadata) {
    schemaMutationLock.runWithLock(() -> {
        createTablesWithLock(tableNamesToTableMetadata));
        internalPutMetadataForTables(tableNamesToTableMetadata);
    }
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/739)*

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java (hunk 1 above)
Comment (clockfort) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java]: Sure, I can probably just give it a slightly more specific name like run...EnsureReturnValuesEqual or something

**Comment by clockfort** [comment]
abandoning, as cause of all of issues already found to just be incorrect version checking

**Pull request closed without merging**